source/agora/script/Signature.d

```diff
@@ -0,0 +1,182 @@
+
+    Decodes the raw byte representation of a signature to its
+    `Signature` and `SigHash` parts, and validates the `SigHash`
+    to be one of the known flags or accepted combination of flags.
+
+    Params:
+        bytes = contains the <Signature, SigHash> tuple
+        sig_pair = will contain the signature tuple if the Signature was encoded
+                   correctly and the SigHash is one of the known flags or
+                   accepted combination of flags.
+
+    Returns:
+        null if the signature tuple was decoded correctly,
+        otherwise the string explaining the reason why if decoding failed
+
+*******************************************************************************/
+
+public string decodeSignature (const(ubyte)[] bytes,
+    out SigPair sig_pair) pure nothrow @safe @nogc
+{
+    if (bytes.length != SigPair.sizeof)
+        return "Encoded signature tuple is of the wrong size";
+
+    const sig = toSignature(bytes[0 .. Signature.sizeof]);
+    bytes.popFrontN(Signature.sizeof);
+
+    assert(bytes.length == 1);
+    const SigHash sig_hash = cast(SigHash)bytes[0];
+    if (!isValidSigHash(sig_hash))
+        return "Unknown SigHash";
+
+    sig_pair = SigPair(sig, sig_hash);
+    return null;
+}
+
+/*******************************************************************************
+
+    Validates that the given `SigHash` is one of the known flags or one of
+    the accepted combination of flags.
+
+    Params:
+        sig_hash = the `SigHash` to validate
+
+    Returns:
+        true if this is one of the known flags or accepted combination of flags
+
+*******************************************************************************/
+
+private bool isValidSigHash (in SigHash sig_hash) pure nothrow @safe @nogc
+{
+    switch (sig_hash)
+    {
+    case SigHash.All:
+        break;
+
+    default:
+        return false;
+    }
+
+    return true;
+}
+
+///
+unittest
+{
+    assert(!isValidSigHash(cast(SigHash)0));
+    assert(isValidSigHash(SigHash.All));
+}
+
+/*******************************************************************************
+
+    Gets the challenge hash for the provided transaction, input index,
+    and the type of SigHash. This cannot be folded into a `sign` routine
+    because it's also required during signature validation.
+
+    The input index is only used for some types of SigHash (SigHash.NoInput).
+
+    Params:
+        tx = the transaction to sign
+        sig_hash = the `SigHash` to use
+        input_idx = the associated input index we're signing for
+
+    Returns:
+        the challenge as a hash
+
+*******************************************************************************/
+
+public Hash getChallenge (in Transaction tx, in SigHash sig_hash = SigHash.All,
+    in ulong input_idx = 0) nothrow @safe
+{
+    assert(input_idx < tx.inputs.length, "Input index is out of range");
+
+    switch (sig_hash)
```

Comment:
You could do a `final switch` here if you ensure that no random byte gets casted to `SigHash` :)